File: ios/HealthTracker/Models/MFPImporter.swift
import Foundation
import CoreData

/// Imports a MyFitnessPal CSV export into the diary. Handles the three export
/// types MyFitnessPal produces, auto-detected from the header row:
///   - Nutrition  -> FoodEntry  (one row per logged food, macros already consumed)
///   - Exercise   -> ExerciseEntry
///   - Measurement -> WeightEntry
///
/// Column lookup is header-name based (case-insensitive, tolerant of the
/// "(g)"/"(mg)" suffixes and column reordering) so small format drift between
/// MFP export versions doesn't break the import.
enum MFPImporter {

    // MARK: - Types

    enum Kind {
        case nutrition, exercise, weight
    }

    /// MFP measurement exports don't reliably label the weight unit, so the user
    /// confirms it before importing. The app stores weight in pounds.
    enum WeightUnit {
        case pounds, kilograms
        var toPounds: Double { self == .kilograms ? 2.2046226218 : 1.0 }
    }

    struct ParsedFoodRow {
        let date: Date
        let mealType: MealType
        let name: String
        let calories, protein, carbs, fat, fiber, sugar, saturatedFat, cholesterol, sodium: Double
        let additional: [String: Double]
    }

    struct ParsedExerciseRow {
        let date: Date
        let name: String
        let category: String
        let durationMinutes: Int
        let caloriesBurned: Double
    }

    struct ParsedWeightRow {
        let date: Date
        let rawWeight: Double
    }

    struct Summary {
        let kind: Kind
        let totalRows: Int
        let importableRows: Int
        let skippedRows: Int
        let startDate: Date?
        let endDate: Date?
        let mealCounts: [MealType: Int]       // nutrition only
        let sampleNames: [String]
        let foodEntries: [ParsedFoodRow]
        let exerciseEntries: [ParsedExerciseRow]
        let weightEntries: [ParsedWeightRow]

        /// Human label for the kind of thing being imported (e.g. "42 entries").
        var noun: String {
            switch kind {
            case .nutrition: return importableRows == 1 ? "entry" : "entries"
            case .exercise: return importableRows == 1 ? "workout" : "workouts"
            case .weight: return importableRows == 1 ? "weigh-in" : "weigh-ins"
            }
        }
    }

    enum ImportError: LocalizedError {
        case empty
        case unrecognized

        var errorDescription: String? {
            switch self {
            case .empty:
                return "That file has no rows to import."
            case .unrecognized:
                return "This doesn't look like a MyFitnessPal export. Choose the Nutrition, Exercise, or Measurement CSV from your MyFitnessPal data."
            }
        }
    }

    // MARK: - Preview (parse only, no insert)

    static func preview(csvText: String) throws -> Summary {
        let (headers, rows) = CSVParser.parseWithHeaders(csvText)
        guard !rows.isEmpty else { throw ImportError.empty }

        let keys = headers.map { $0.lowercased() }
        guard let kind = detectKind(keys) else { throw ImportError.unrecognized }

        switch kind {
        case .nutrition: return previewNutrition(rows)
        case .exercise: return previewExercise(rows)
        case .weight: return previewWeight(rows)
        }
    }

    private static func detectKind(_ keys: [String]) -> Kind? {
        func has(_ prefix: String) -> Bool { keys.contains { $0.hasPrefix(prefix) } }
        // Order matters: a strength row in an exercise export can carry a "weight"
        // column, so check food and exercise before falling back to weight.
        if has("food") { return .nutrition }
        if has("exercise") { return .exercise }
        if has("weight") { return .weight }
        return nil
    }

    // MARK: - Import (batched insert with dedupe)

    /// Inserts the summary's rows and returns how many were actually added
    /// (duplicates already in the diary are skipped). Dedupe is against entries
    /// that existed *before* this import, so a full re-import adds nothing.
    @discardableResult
    static func importSummary(
        _ summary: Summary,
        weightUnit: WeightUnit,
        container: NSPersistentContainer,
        progress: @escaping (Double) -> Void
    ) async throws -> Int {
        switch summary.kind {
        case .nutrition:
            return try await insert(summary.foodEntries, container: container, progress: progress)
        case .exercise:
            return try await insert(summary.exerciseEntries, container: container, progress: progress)
        case .weight:
            return try await insert(summary.weightEntries, unit: weightUnit, container: container, progress: progress)
        }
    }

    // MARK: - Nutrition

    private static func previewNutrition(_ rows: [[String: String]]) -> Summary {
        var parsed: [ParsedFoodRow] = []
        var skipped = 0
        for row in rows {
            if let e = parseFood(row) { parsed.append(e) } else { skipped += 1 }
        }
        let sorted = parsed.sorted { $0.date < $1.date }
        var mealCounts: [MealType: Int] = [:]
        for e in parsed { mealCounts[e.mealType, default: 0] += 1 }
        return Summary(
            kind: .nutrition, totalRows: rows.count, importableRows: parsed.count, skippedRows: skipped,
            startDate: sorted.first?.date, endDate: sorted.last?.date, mealCounts: mealCounts,
            sampleNames: Array(parsed.prefix(5).map { $0.name }),
            foodEntries: parsed, exerciseEntries: [], weightEntries: []
        )
    }

    private static func insert(_ entries: [ParsedFoodRow], container: NSPersistentContainer, progress: @escaping (Double) -> Void) async throws -> Int {
        guard !entries.isEmpty else { return 0 }
        let context = container.newBackgroundContext()
        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
        return try await context.perform {
            let existing = try existingKeys(in: context, entity: "FoodEntry", dates: entries.map { $0.date }) { e in
                dedupeKey(day: e.value(forKey: "date") as? Date, name: e.value(forKey: "name") as? String ?? "", n: e.value(forKey: "calories") as? Double ?? 0)
            }
            var inserted = 0
            let total = entries.count
            for (idx, row) in entries.enumerated() {
                let key = dedupeKey(day: row.date, name: row.name, n: row.calories)
                if existing.contains(key) { continue }
                let entry = FoodEntry(context: context)
                entry.id = UUID()
                entry.name = row.name
                entry.mealType = row.mealType.rawValue
                entry.date = row.date
                entry.timestamp = row.date
                entry.calories = row.calories
                entry.protein = row.protein
                entry.carbs = row.carbs
                entry.fat = row.fat
                entry.fiber = row.fiber
                entry.sugar = row.sugar
                entry.saturatedFat = row.saturatedFat
                entry.cholesterol = row.cholesterol
                entry.sodium = row.sodium
                entry.servingSize = "1"
                entry.servingUnit = "serving"
                entry.servingCount = 1
                if !row.additional.isEmpty { entry.additionalNutrients = row.additional }
                inserted += 1
                if inserted % 500 == 0 { try context.save(); context.reset() }
                if idx % 100 == 0 { progress(Double(idx) / Double(total)) }
            }
            if context.hasChanges { try context.save() }
            progress(1.0)
            return inserted
        }
    }

    private static func parseFood(_ row: [String: String]) -> ParsedFoodRow? {
        guard let dateStr = value(row, ["date"]), let date = parseDate(dateStr) else { return nil }
        guard let name = value(row, ["food", "name"]), !name.isEmpty else { return nil }
        let meal = mapMeal(value(row, ["meal"]) ?? "")
        let dated = timestamp(for: meal, on: date)
        var additional: [String: Double] = [:]
        if let p = number(row, ["potassium"]) { additional["Potassium"] = p }
        if let v = number(row, ["vitamin a"]) { additional["Vitamin A"] = v }
        if let v = number(row, ["vitamin c"]) { additional["Vitamin C"] = v }
        if let v = number(row, ["calcium"]) { additional["Calcium"] = v }
        if let v = number(row, ["iron"]) { additional["Iron"] = v }
        return ParsedFoodRow(
            date: dated, mealType: meal, name: name,
            calories: number(row, ["calories", "energy"]) ?? 0,
            protein: number(row, ["protein (g)", "protein"]) ?? 0,
            carbs: number(row, ["carbohydrates (g)", "carbohydrates", "carbs"]) ?? 0,
            fat: number(row, ["fat (g)", "fat"]) ?? 0,
            fiber: number(row, ["fiber", "fibre"]) ?? 0,
            sugar: number(row, ["sugar"]) ?? 0,
            saturatedFat: number(row, ["saturated fat"]) ?? 0,
            cholesterol: number(row, ["cholesterol"]) ?? 0,
            sodium: number(row, ["sodium (mg)", "sodium"]) ?? 0,
            additional: additional
        )
    }

    // MARK: - Exercise

    private static func previewExercise(_ rows: [[String: String]]) -> Summary {
        var parsed: [ParsedExerciseRow] = []
        var skipped = 0
        for row in rows {
            if let e = parseExercise(row) { parsed.append(e) } else { skipped += 1 }
        }
        let sorted = parsed.sorted { $0.date < $1.date }
        return Summary(
            kind: .exercise, totalRows: rows.count, importableRows: parsed.count, skippedRows: skipped,
            startDate: sorted.first?.date, endDate: sorted.last?.date, mealCounts: [:],
            sampleNames: Array(parsed.prefix(5).map { $0.name }),
            foodEntries: [], exerciseEntries: parsed, weightEntries: []
        )
    }

    private static func insert(_ entries: [ParsedExerciseRow], container: NSPersistentContainer, progress: @escaping (Double) -> Void) async throws -> Int {
        guard !entries.isEmpty else { return 0 }
        let context = container.newBackgroundContext()
        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
        return try await context.perform {
            let existing = try existingKeys(in: context, entity: "ExerciseEntry", dates: entries.map { $0.date }) { e in
                dedupeKey(day: e.value(forKey: "date") as? Date, name: e.value(forKey: "name") as? String ?? "", n: e.value(forKey: "caloriesBurned") as? Double ?? 0)
            }
            var inserted = 0
            let total = entries.count
            for (idx, row) in entries.enumerated() {
                let key = dedupeKey(day: row.date, name: row.name, n: row.caloriesBurned)
                if existing.contains(key) { continue }
                let entry = ExerciseEntry(context: context)
                entry.id = UUID()
                entry.name = row.name
                entry.category = row.category
                entry.type = row.category
                entry.duration = Int32(row.durationMinutes)
                entry.caloriesBurned = row.caloriesBurned
                entry.date = row.date
                entry.timestamp = row.date
                inserted += 1
                if inserted % 500 == 0 { try context.save(); context.reset() }
                if idx % 100 == 0 { progress(Double(idx) / Double(total)) }
            }
            if context.hasChanges { try context.save() }
            progress(1.0)
            return inserted
        }
    }

    private static func parseExercise(_ row: [String: String]) -> ParsedExerciseRow? {
        guard let dateStr = value(row, ["date"]), let date = parseDate(dateStr) else { return nil }
        guard let name = value(row, ["exercise", "name"]), !name.isEmpty else { return nil }
        let type = value(row, ["type"]) ?? "Cardio"
        let category = mapExerciseCategory(type)
        let minutes = Int(number(row, ["exercise minutes", "minutes", "duration"]) ?? 0)
        let calories = number(row, ["exercise calories", "calories burned", "calories"]) ?? 0
        // Assign a stable midday time so imported workouts order sensibly.
        let dated = Calendar.current.date(bySettingHour: 12, minute: 0, second: 0, of: date) ?? date
        return ParsedExerciseRow(date: dated, name: name, category: category, durationMinutes: minutes, caloriesBurned: calories)
    }

    private static func mapExerciseCategory(_ raw: String) -> String {
        let lower = raw.lowercased()
        if lower.contains("strength") || lower.contains("weight") { return "Strength" }
        if lower.contains("cardio") { return "Cardio" }
        return raw.isEmpty ? "Cardio" : raw
    }

    // MARK: - Weight

    private static func previewWeight(_ rows: [[String: String]]) -> Summary {
        var parsed: [ParsedWeightRow] = []
        var skipped = 0
        for row in rows {
            if let e = parseWeight(row) { parsed.append(e) } else { skipped += 1 }
        }
        let sorted = parsed.sorted { $0.date < $1.date }
        return Summary(
            kind: .weight, totalRows: rows.count, importableRows: parsed.count, skippedRows: skipped,
            startDate: sorted.first?.date, endDate: sorted.last?.date, mealCounts: [:],
            sampleNames: [], foodEntries: [], exerciseEntries: [], weightEntries: parsed
        )
    }

    private static func insert(_ entries: [ParsedWeightRow], unit: WeightUnit, container: NSPersistentContainer, progress: @escaping (Double) -> Void) async throws -> Int {
        guard !entries.isEmpty else { return 0 }
        let factor = unit.toPounds
        let context = container.newBackgroundContext()
        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
        return try await context.perform {
            let existing = try existingKeys(in: context, entity: "WeightEntry", dates: entries.map { $0.date }) { e in
                weightKey(day: e.value(forKey: "date") as? Date ?? e.value(forKey: "timestamp") as? Date, pounds: e.value(forKey: "weight") as? Double ?? 0)
            }
            var inserted = 0
            let total = entries.count
            for (idx, row) in entries.enumerated() {
                let pounds = row.rawWeight * factor
                let key = weightKey(day: row.date, pounds: pounds)
                if existing.contains(key) { continue }
                let entry = WeightEntry(context: context)
                entry.id = UUID()
                entry.weight = pounds
                entry.date = row.date
                entry.timestamp = row.date
                inserted += 1
                if idx % 100 == 0 { progress(Double(idx) / Double(total)) }
            }
            if context.hasChanges { try context.save() }
            progress(1.0)
            return inserted
        }
    }

    private static func parseWeight(_ row: [String: String]) -> ParsedWeightRow? {
        guard let dateStr = value(row, ["date"]), let date = parseDate(dateStr) else { return nil }
        guard let w = number(row, ["weight"]), w > 0 else { return nil }
        let dated = Calendar.current.date(bySettingHour: 7, minute: 0, second: 0, of: date) ?? date
        return ParsedWeightRow(date: dated, rawWeight: w)
    }

    // MARK: - Meal + time mapping

    private static func mapMeal(_ raw: String) -> MealType {
        switch raw.trimmingCharacters(in: .whitespaces).lowercased() {
        case "breakfast": return .breakfast
        case "lunch": return .lunch
        case "dinner": return .dinner
        case "snack", "snacks": return .snack
        default: return .snack
        }
    }

    private static func timestamp(for meal: MealType, on date: Date) -> Date {
        let cal = Calendar.current
        let (h, m): (Int, Int)
        switch meal {
        case .breakfast: (h, m) = (8, 0)
        case .lunch: (h, m) = (12, 30)
        case .dinner: (h, m) = (19, 0)
        case .snack: (h, m) = (15, 0)
        }
        return cal.date(bySettingHour: h, minute: m, second: 0, of: date) ?? date
    }

    // MARK: - Dedupe

    private static func dedupeKey(day: Date?, name: String, n: Double) -> String {
        let dayStr = (day.map { Calendar.current.startOfDay(for: $0).timeIntervalSince1970 } ?? 0)
        return "\(dayStr)|\(name.lowercased())|\(Int(n.rounded()))"
    }

    private static func weightKey(day: Date?, pounds: Double) -> String {
        let dayStr = (day.map { Calendar.current.startOfDay(for: $0).timeIntervalSince1970 } ?? 0)
        return "\(dayStr)|\(Int((pounds * 10).rounded()))"   // 0.1 lb resolution
    }

    private static func existingKeys(in context: NSManagedObjectContext, entity: String, dates: [Date], key: (NSManagedObject) -> String) throws -> Set<String> {
        guard let minDate = dates.min(), let maxDate = dates.max() else { return [] }
        let cal = Calendar.current
        let start = cal.startOfDay(for: minDate)
        let end = cal.date(byAdding: .day, value: 1, to: cal.startOfDay(for: maxDate)) ?? maxDate
        let request = NSFetchRequest<NSManagedObject>(entityName: entity)
        request.predicate = NSPredicate(format: "date >= %@ AND date < %@", start as NSDate, end as NSDate)
        let existing = try context.fetch(request)
        return Set(existing.map(key))
    }

    // MARK: - Column helpers

    private static func value(_ row: [String: String], _ candidates: [String]) -> String? {
        for candidate in candidates {
            if let exact = row[candidate], !exact.isEmpty { return exact.trimmingCharacters(in: .whitespaces) }
        }
        for candidate in candidates {
            if let match = row.first(where: { $0.key.hasPrefix(candidate) })?.value, !match.isEmpty {
                return match.trimmingCharacters(in: .whitespaces)
            }
        }
        return nil
    }

    private static func number(_ row: [String: String], _ candidates: [String]) -> Double? {
        guard let raw = value(row, candidates) else { return nil }
        let cleaned = raw.replacingOccurrences(of: ",", with: "").replacingOccurrences(of: " ", with: "")
        return Double(cleaned)
    }

    private static func parseDate(_ raw: String) -> Date? {
        let trimmed = raw.trimmingCharacters(in: .whitespaces)
        for format in dateFormats {
            let df = DateFormatter()
            df.locale = Locale(identifier: "en_US_POSIX")
            df.dateFormat = format
            if let d = df.date(from: trimmed) { return d }
        }
        return nil
    }

    private static let dateFormats = [
        "yyyy-MM-dd", "MM/dd/yyyy", "M/d/yyyy", "dd/MM/yyyy", "d/M/yyyy", "yyyy/MM/dd", "MMMM d, yyyy"
    ]
}

File: android/app/src/main/java/com/mochasmindlab/mlhealth/data/importer/MfpImporter.kt
package com.mochasmindlab.mlhealth.data.importer

import com.mochasmindlab.mlhealth.data.database.MLFitnessDatabase
import com.mochasmindlab.mlhealth.data.entities.ExerciseEntry
import com.mochasmindlab.mlhealth.data.entities.FoodEntry
import com.mochasmindlab.mlhealth.data.entities.WeightEntry
import com.mochasmindlab.mlhealth.data.models.MealType
import com.mochasmindlab.mlhealth.util.CsvParser
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import kotlin.math.roundToInt

/**
 * Imports a MyFitnessPal CSV export into the diary. Handles the three export
 * types MyFitnessPal produces, auto-detected from the header row:
 *   - Nutrition   -> FoodEntry  (one row per logged food, macros already consumed)
 *   - Exercise    -> ExerciseEntry
 *   - Measurement -> WeightEntry
 *
 * Column lookup is header-name based (case-insensitive, tolerant of "(g)"/"(mg)"
 * suffixes and column reordering). Mirrors the iOS MFPImporter.
 */
object MfpImporter {

    enum class Kind { NUTRITION, EXERCISE, WEIGHT }

    /** The app stores weight in pounds; MFP measurement exports don't reliably
     * label the unit, so the user confirms it before importing. */
    enum class WeightUnit(val toPounds: Double) {
        POUNDS(1.0), KILOGRAMS(2.2046226218)
    }

    data class ParsedFoodRow(
        val date: Date, val mealType: MealType, val name: String,
        val calories: Double, val protein: Double, val carbs: Double, val fat: Double,
        val fiber: Double, val sugar: Double, val saturatedFat: Double,
        val cholesterol: Double, val sodium: Double, val additional: Map<String, Double>
    )

    data class ParsedExerciseRow(
        val date: Date, val name: String, val category: String,
        val durationMinutes: Int, val caloriesBurned: Double
    )

    data class ParsedWeightRow(val date: Date, val rawWeight: Double)

    data class Summary(
        val kind: Kind,
        val totalRows: Int,
        val importableRows: Int,
        val skippedRows: Int,
        val startDate: Date?,
        val endDate: Date?,
        val mealCounts: Map<MealType, Int>,
        val sampleNames: List<String>,
        val foodEntries: List<ParsedFoodRow>,
        val exerciseEntries: List<ParsedExerciseRow>,
        val weightEntries: List<ParsedWeightRow>
    ) {
        val noun: String
            get() = when (kind) {
                Kind.NUTRITION -> if (importableRows == 1) "entry" else "entries"
                Kind.EXERCISE -> if (importableRows == 1) "workout" else "workouts"
                Kind.WEIGHT -> if (importableRows == 1) "weigh-in" else "weigh-ins"
            }
    }

    class ImportException(message: String) : Exception(message)

    // MARK: - Preview

    fun preview(csvText: String): Summary {
        val (headers, rows) = CsvParser.parseWithHeaders(csvText)
        if (rows.isEmpty()) throw ImportException("That file has no rows to import.")

        val keys = headers.map { it.lowercase() }
        return when (detectKind(keys)) {
            Kind.NUTRITION -> previewNutrition(rows)
            Kind.EXERCISE -> previewExercise(rows)
            Kind.WEIGHT -> previewWeight(rows)
            null -> throw ImportException(
                "This doesn't look like a MyFitnessPal export. Choose the Nutrition, Exercise, or Measurement CSV from your MyFitnessPal data."
            )
        }
    }

    private fun detectKind(keys: List<String>): Kind? {
        fun has(prefix: String) = keys.any { it.startsWith(prefix) }
        // Order matters: a strength row in an exercise export can carry a
        // "weight" column, so check food and exercise before weight.
        return when {
            has("food") -> Kind.NUTRITION
            has("exercise") -> Kind.EXERCISE
            has("weight") -> Kind.WEIGHT
            else -> null
        }
    }

    // MARK: - Import

    /**
     * Inserts the summary's rows and returns how many were actually added
     * (duplicates already in the diary are skipped). Dedupe is against entries
     * that existed *before* this import, so a full re-import adds nothing.
     */
    suspend fun importSummary(
        summary: Summary,
        weightUnit: WeightUnit,
        db: MLFitnessDatabase,
        progress: (Double) -> Unit
    ): Int = when (summary.kind) {
        Kind.NUTRITION -> insertFood(summary.foodEntries, db, progress)
        Kind.EXERCISE -> insertExercise(summary.exerciseEntries, db, progress)
        Kind.WEIGHT -> insertWeight(summary.weightEntries, weightUnit, db, progress)
    }

    // MARK: - Nutrition

    private fun previewNutrition(rows: List<Map<String, String>>): Summary {
        val parsed = mutableListOf<ParsedFoodRow>()
        var skipped = 0
        for (row in rows) parseFood(row)?.let { parsed.add(it) } ?: skipped++
        val sorted = parsed.sortedBy { it.date }
        return Summary(
            Kind.NUTRITION, rows.size, parsed.size, skipped,
            sorted.firstOrNull()?.date, sorted.lastOrNull()?.date,
            parsed.groupingBy { it.mealType }.eachCount(),
            parsed.take(5).map { it.name }, parsed, emptyList(), emptyList()
        )
    }

    private suspend fun insertFood(entries: List<ParsedFoodRow>, db: MLFitnessDatabase, progress: (Double) -> Unit): Int {
        if (entries.isEmpty()) return 0
        val dao = db.foodDao()
        val (start, end) = dayRange(entries.map { it.date })
        val existing = dao.getEntriesBetween(start, end)
            .map { dedupeKey(it.date, it.name, it.calories) }.toSet()
        var inserted = 0
        val total = entries.size
        entries.forEachIndexed { idx, row ->
            if (!existing.contains(dedupeKey(row.date, row.name, row.calories))) {
                dao.insert(
                    FoodEntry(
                        name = row.name, date = row.date, timestamp = row.date,
                        mealType = row.mealType.name.lowercase(),
                        servingSize = "1", servingUnit = "serving", servingCount = 1.0,
                        calories = row.calories, protein = row.protein, carbs = row.carbs, fat = row.fat,
                        fiber = row.fiber, sugar = row.sugar, sodium = row.sodium,
                        cholesterol = row.cholesterol, saturatedFat = row.saturatedFat,
                        additionalNutrients = row.additional
                    )
                )
                inserted++
            }
            if (idx % 100 == 0) progress(idx.toDouble() / total)
        }
        progress(1.0)
        return inserted
    }

    private fun parseFood(row: Map<String, String>): ParsedFoodRow? {
        val date = value(row, listOf("date"))?.let { parseDate(it) } ?: return null
        val name = value(row, listOf("food", "name"))?.takeIf { it.isNotEmpty() } ?: return null
        val meal = mapMeal(value(row, listOf("meal")) ?: "")
        val dated = timestamp(meal, date)
        val additional = mutableMapOf<String, Double>()
        number(row, listOf("potassium"))?.let { additional["Potassium"] = it }
        number(row, listOf("vitamin a"))?.let { additional["Vitamin A"] = it }
        number(row, listOf("vitamin c"))?.let { additional["Vitamin C"] = it }
        number(row, listOf("calcium"))?.let { additional["Calcium"] = it }
        number(row, listOf("iron"))?.let { additional["Iron"] = it }
        return ParsedFoodRow(
            date = dated, mealType = meal, name = name,
            calories = number(row, listOf("calories", "energy")) ?: 0.0,
            protein = number(row, listOf("protein (g)", "protein")) ?: 0.0,
            carbs = number(row, listOf("carbohydrates (g)", "carbohydrates", "carbs")) ?: 0.0,
            fat = number(row, listOf("fat (g)", "fat")) ?: 0.0,
            fiber = number(row, listOf("fiber", "fibre")) ?: 0.0,
            sugar = number(row, listOf("sugar")) ?: 0.0,
            saturatedFat = number(row, listOf("saturated fat")) ?: 0.0,
            cholesterol = number(row, listOf("cholesterol")) ?: 0.0,
            sodium = number(row, listOf("sodium (mg)", "sodium")) ?: 0.0,
            additional = additional
        )
    }

    // MARK: - Exercise

    private fun previewExercise(rows: List<Map<String, String>>): Summary {
        val parsed = mutableListOf<ParsedExerciseRow>()
        var skipped = 0
        for (row in rows) parseExercise(row)?.let { parsed.add(it) } ?: skipped++
        val sorted = parsed.sortedBy { it.date }
        return Summary(
            Kind.EXERCISE, rows.size, parsed.size, skipped,
            sorted.firstOrNull()?.date, sorted.lastOrNull()?.date, emptyMap(),
            parsed.take(5).map { it.name }, emptyList(), parsed, emptyList()
        )
    }

    private suspend fun insertExercise(entries: List<ParsedExerciseRow>, db: MLFitnessDatabase, progress: (Double) -> Unit): Int {
        if (entries.isEmpty()) return 0
        val dao = db.exerciseDao()
        val (start, end) = dayRange(entries.map { it.date })
        val existing = dao.getExercisesBetweenOnce(start, end)
            .map { dedupeKey(it.date, it.name, it.caloriesBurned) }.toSet()
        var inserted = 0
        val total = entries.size
        entries.forEachIndexed { idx, row ->
            if (!existing.contains(dedupeKey(row.date, row.name, row.caloriesBurned))) {
                dao.insert(
                    ExerciseEntry(
                        name = row.name, category = row.category, type = row.category,
                        date = row.date, timestamp = row.date,
                        duration = row.durationMinutes, caloriesBurned = row.caloriesBurned
                    )
                )
                inserted++
            }
            if (idx % 100 == 0) progress(idx.toDouble() / total)
        }
        progress(1.0)
        return inserted
    }

    private fun parseExercise(row: Map<String, String>): ParsedExerciseRow? {
        val date = value(row, listOf("date"))?.let { parseDate(it) } ?: return null
        val name = value(row, listOf("exercise", "name"))?.takeIf { it.isNotEmpty() } ?: return null
        val category = mapExerciseCategory(value(row, listOf("type")) ?: "Cardio")
        val minutes = (number(row, listOf("exercise minutes", "minutes", "duration")) ?: 0.0).toInt()
        val calories = number(row, listOf("exercise calories", "calories burned", "calories")) ?: 0.0
        val dated = atTime(date, 12, 0)
        return ParsedExerciseRow(dated, name, category, minutes, calories)
    }

    private fun mapExerciseCategory(raw: String): String {
        val lower = raw.lowercase()
        return when {
            lower.contains("strength") || lower.contains("weight") -> "Strength"
            lower.contains("cardio") -> "Cardio"
            raw.isEmpty() -> "Cardio"
            else -> raw
        }
    }

    // MARK: - Weight

    private fun previewWeight(rows: List<Map<String, String>>): Summary {
        val parsed = mutableListOf<ParsedWeightRow>()
        var skipped = 0
        for (row in rows) parseWeight(row)?.let { parsed.add(it) } ?: skipped++
        val sorted = parsed.sortedBy { it.date }
        return Summary(
            Kind.WEIGHT, rows.size, parsed.size, skipped,
            sorted.firstOrNull()?.date, sorted.lastOrNull()?.date, emptyMap(),
            emptyList(), emptyList(), emptyList(), parsed
        )
    }

    private suspend fun insertWeight(entries: List<ParsedWeightRow>, unit: WeightUnit, db: MLFitnessDatabase, progress: (Double) -> Unit): Int {
        if (entries.isEmpty()) return 0
        val dao = db.weightDao()
        val factor = unit.toPounds
        val (start, end) = dayRange(entries.map { it.date })
        val existing = dao.getEntriesInRange(start, end)
            .map { weightKey(it.date, it.weight) }.toSet()
        var inserted = 0
        val total = entries.size
        entries.forEachIndexed { idx, row ->
            val pounds = row.rawWeight * factor
            if (!existing.contains(weightKey(row.date, pounds))) {
                dao.insert(WeightEntry(weight = pounds, date = row.date, timestamp = row.date))
                inserted++
            }
            if (idx % 100 == 0) progress(idx.toDouble() / total)
        }
        progress(1.0)
        return inserted
    }

    private fun parseWeight(row: Map<String, String>): ParsedWeightRow? {
        val date = value(row, listOf("date"))?.let { parseDate(it) } ?: return null
        val w = number(row, listOf("weight"))?.takeIf { it > 0 } ?: return null
        return ParsedWeightRow(atTime(date, 7, 0), w)
    }

    // MARK: - Meal + time mapping

    private fun mapMeal(raw: String): MealType = when (raw.trim().lowercase()) {
        "breakfast" -> MealType.BREAKFAST
        "lunch" -> MealType.LUNCH
        "dinner" -> MealType.DINNER
        "snack", "snacks" -> MealType.SNACK
        else -> MealType.SNACK
    }

    private fun timestamp(meal: MealType, date: Date): Date {
        val (h, m) = when (meal) {
            MealType.BREAKFAST -> 8 to 0
            MealType.LUNCH -> 12 to 30
            MealType.DINNER -> 19 to 0
            MealType.SNACK -> 15 to 0
        }
        return atTime(date, h, m)
    }

    private fun atTime(date: Date, hour: Int, minute: Int): Date {
        val cal = Calendar.getInstance()
        cal.time = date
        cal.set(Calendar.HOUR_OF_DAY, hour)
        cal.set(Calendar.MINUTE, minute)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    // MARK: - Dedupe

    private fun startOfDay(date: Date): Date {
        val cal = Calendar.getInstance()
        cal.time = date
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    private fun dayRange(dates: List<Date>): Pair<Date, Date> {
        val min = dates.minOrNull() ?: Date(0)
        val max = dates.maxOrNull() ?: Date(0)
        val start = startOfDay(min)
        val cal = Calendar.getInstance()
        cal.time = startOfDay(max)
        cal.add(Calendar.DAY_OF_YEAR, 1)
        return Pair(start, cal.time)
    }

    private fun dedupeKey(day: Date, name: String, n: Double): String =
        "${startOfDay(day).time}|${name.lowercase()}|${n.roundToInt()}"

    private fun weightKey(day: Date, pounds: Double): String =
        "${startOfDay(day).time}|${(pounds * 10).roundToInt()}"   // 0.1 lb resolution

    // MARK: - Column helpers

    private fun value(row: Map<String, String>, candidates: List<String>): String? {
        for (candidate in candidates) {
            val exact = row[candidate]
            if (!exact.isNullOrEmpty()) return exact.trim()
        }
        for (candidate in candidates) {
            val match = row.entries.firstOrNull { it.key.startsWith(candidate) }?.value
            if (!match.isNullOrEmpty()) return match.trim()
        }
        return null
    }

    private fun number(row: Map<String, String>, candidates: List<String>): Double? {
        val raw = value(row, candidates) ?: return null
        return raw.replace(",", "").replace(" ", "").toDoubleOrNull()
    }

    private fun parseDate(raw: String): Date? {
        val trimmed = raw.trim()
        for (format in dateFormats) {
            try {
                val df = SimpleDateFormat(format, Locale.US)
                df.isLenient = false
                return df.parse(trimmed) ?: continue
            } catch (_: Exception) {
            }
        }
        return null
    }

    private val dateFormats = listOf(
        "yyyy-MM-dd", "MM/dd/yyyy", "M/d/yyyy", "dd/MM/yyyy", "d/M/yyyy", "yyyy/MM/dd", "MMMM d, yyyy"
    )
}
